[jiayr99] iP

jiayr99 opened the pull request:
# SageBot
> "The best way to predict the future is to invent it" - Allan Kay [(source)](https://www.goodreads.com/author/quotes/304595.Alan_Kay#:~:text=The%20best%20way%20to%20predict%20the%20future%20is%20to%20invent%20it.&text=The%20most%20disastrous%20thing%20that,is%20your%20first%20programming%20language.&text=A%20change%20in%20perspective%20is%20worth%2080%20IQ%20points.&text=Technology%20is%20anything%20invented%20after%20you%20were%20born.)

Tired of remembering things to do? Look no further! SageBot is,
- text-based
* easy to learn
+ ~FAST~ ***SUPER FAST*** to use

Follow these easy steps to download the application:
1. download it from [here](https://github.com/jiayr99/ip)
2. double click it
3. add your tasks
4. let it manage your tasks for you :stuck_out_tongue_closed_eyes:

And the best part is it is **FREE!!**

Features:
- [X] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders

If you are a Java programmer, you can use it to practise Java too. Here's the `main` method:
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

jiayr99 committed:
Rename main class

jiayr99 committed:
Add greeting and exit

jiayr99 committed:
Added Echo

jiayr99 committed:
create an array list to store user inputs

jiayr99 committed:
Display input back to user

jiayr99 committed:
Create Task class

jiayr99 committed:
Change list to store tasks

jiayr99 committed:
Mark tasks as done and unmark tasks as not done

jiayr99 committed:
Create Deadline class

jiayr99 committed:
Create ToDo class

jiayr99 committed:
Create Event class

jiayr99 committed:
Support ToDo

jiayr99 committed:
Support Deadline

jiayr99 committed:
Support event

jiayr99 committed:
Match name of my main class

jiayr99 committed:
Added input text

jiayr99 committed:
Updated expected text

jiayr99 committed:
Create an exception class

jiayr99 committed:
Add exception in ToDo class

jiayr99 committed:
Adding try and catch block

jiayr99 committed:
Add exception to deadline

jiayr99 committed:
Add exception to event

jiayr99 committed:
Handle all the exceptions in todo, deadline and event

jiayr99 committed:
Changed to an array list of tasks

jiayr99 committed:
Added the delete function

jiayr99 committed:
change name of bot

jiayr99 committed:
Update name of bot

jiayr99 committed:
Update exception to new name of bot

jiayr99 committed:
Create storage class to read and write from the file

jiayr99 committed:
Add method toFileFormat()

jiayr99 committed:
Update main file to save tasks and load tasks

jiayr99 committed:
Debug error in todo

jiayr99 committed:
Add header comments to methods

jiayr99 committed:
Merge branch 'branch-level-7'

jiayr99 committed:
Update Deadline class to store the actual dates

jiayr99 committed:
Update Event class to store the correct date and time format

jiayr99 committed:
Add getters to Deadline and Event class

jiayr99 committed:
Add command for "on" to print tasks on certain dates

jiayr99 committed:
Create Ui class to deal with interactions with user

jiayr99 committed:
Create TaskList class to contain the task list

jiayr99 committed:
Create Parser class to deal with making sense of the user command

jiayr99 committed:
Create absract Command class

jiayr99 committed:
Create classes for all the commands which inherit from Command class

jiayr99 committed:
Update storage class with the new classes created

jiayr99 committed:
Update main class code to be easier to read

jiayr99 committed:
Merge branch 'branch-A-MoreOOP'

jiayr99 committed:
Create package exception

jiayr99 committed:
Create sage package

jiayr99 committed:
Create command package

jiayr99 committed:
Create parser package

jiayr99 committed:
Create storage package

jiayr99 committed:
Create task package

jiayr99 committed:
Merge branch 'branch-A-Packages'

jiayr99 committed:
Merge branch 'add-gradle-support'

jiayr99 committed:
Update main class

jiayr99 committed:
Merge branch 'branch-A-Gradle'

jiayr99 committed:
Add test cases

jiayr99 committed:
Merge branch 'branch-A-JUnit'

jiayr99 committed:
Add header comments to classes and methods

jiayr99 committed:
Add a class for the find command

jiayr99 committed:
Add a find command for the parser

jiayr99 committed:
Add the method to find tasks with matching keyword

jiayr99 committed:
Add method to show line

jiayr99 committed:
Merge branch 'branch-level-9'

jiayr99 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/sage/ui/Ui.java

ttzuche commented on src/test/java/sage/command/CommandTest.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @Test
+    public void testExecute() {
```
The logic in the `testExecute` method is solid. You have simulated adding a `ToDo` task into a mock TaskList and validated the result with an assertion.

ttzuche replied:
Perhaps you could give a more specific name to `testExecute` like `testExecute_addsTaskToTaskList`?

ttzuche commented on src/test/java/sage/command/CommandTest.java:
```diff
@@ -0,0 +1,53 @@
+    @Test
+    public void testIsExit_defaultBehaviour() {
+        // Create an anonymous subclass of command
+        Command command = new Command() {
```
Good use of anonymous subclass of `Command` in both tests!

ttzuche commented on src/test/java/sage/storage/StorageTest.java:
```diff
@@ -0,0 +1,79 @@
+                "The error message should indicate a problem when loading the file");
+    }
+    */
+}
```
These tests cases are great! But it might be beneficial in the future to add more test cases for scenarios when `Storage` fails.

ttzuche reviewed (commented):
Overall code looks really good! However, it would be beneficial to add a few more test cases to ensure comprehensive coverage.

ttzuche commented on src/main/java/sage/Sage.java:
```diff
@@ -9,6 +9,11 @@
+ * The main class for the Sage task management application
+ * This class initializes the application components, manages user interactions,
+ * and controls the overall flow of the application
+ */
```
I like how the javadocs are clear and well-written, which makes understanding the code a whole easier.

ttzuche commented on src/main/java/sage/command/DeleteCommand.java:
```diff
@@ -6,9 +6,18 @@
+     * @param index The 1-based index of the task to be deleted in the task list
+     *              The index is decremented by 1 to match the 0-based indexing of the TaskList
+     */
     public DeleteCommand(int index) {
```
While you did clearly explain the 1-based index in your javadocs, I think that the `index` parameter can be changed to `taskNumber` for better understanding.

ttzuche reviewed (commented):
All javadocs are very well-written!! 👍

jiayr99 committed:
Implement GUI

jiayr99 committed:
Merge branch 'branch-level-10'

seantham21 commented on src/main/java/sage/MainWindow.java:
```diff
@@ -0,0 +1,84 @@
+
+    /**
+     * Creates two dialog boxes, one echoing user input and the other containing Duke's reply and then appends them to
+     * the dialog container. Clears the user input after processing.
```
Perhaps each sentence can start on a new line?

seantham21 commented on src/main/java/sage/MainWindow.java:
```diff
@@ -0,0 +1,84 @@
+     * the dialog container. Clears the user input after processing.
+     */
+    @FXML
+    private void handleUserInput() {
```
While I like how the JavaDocs helps to explain what this method is doing, perhaps a slightly more specific method name could be used for easier understanding on first glance?

seantham21 commented on src/main/java/sage/Main.java:
```diff
@@ -0,0 +1,33 @@
+
+import javafx.application.Application;
+import javafx.fxml.FXMLLoader;
+import javafx.scene.Scene;
+import javafx.scene.layout.AnchorPane;
+import javafx.stage.Stage;
```
I like how the imports are mostly very specific and follow an alphabetical order, but perhaps a few of these slight switches can be made to ensure the whole repo follows the same standard? (Just a nit!)

seantham21 commented on src/main/java/sage/Sage.java:
```diff
@@ -0,0 +1,221 @@
+        }
+    }
+
+    public String getResponse(String input) {
+        try {
+            commandType = Parser.parseCommand(input);
+            switch (commandType) {
+                case BYE :
+                    this.exit();
+                    return this.ui.showGoodbyeMessage();
+                case LIST:
+                    return this.ui.showTaskList(tasks);
+                case MARK:
+                    return markCommand(Parser.parseArgs(CommandType.MARK, input));
+                case UNMARK:
+                    return unmarkCommand(Parser.parseArgs(CommandType.UNMARK, input));
+                case DELETE:
+                    return deleteCommand(Parser.parseArgs(CommandType.DELETE, input));
+                case FIND:
+                    return findCommand(Parser.parseArgs(CommandType.FIND, input));
+                case TODO:
+                    return toDoCommand(Parser.parseArgs(CommandType.TODO, input));
+                case DEADLINE:
+                    return deadlineCommand(Parser.parseArgs(CommandType.DEADLINE, input));
+                case EVENT:
+                    return eventCommand(Parser.parseArgs(CommandType.EVENT, input));
+                default:
+                    throw new SageException("Sorry, what do you mean? :p");
+            }
+        } catch (SageException e) {
+            return this.ui.showErrorMessage(e);
+        }
+    }
```
Personally, I think that some of these might be classified under the Parser class rather than the main class, but that could just be something of personal preference that you could consider!

seantham21 commented on src/main/java/sage/Sage.java:
```diff
@@ -0,0 +1,221 @@
+        }
+    }
+
+    private String eventCommand(String input) throws SageException {
```
I like the naming conventions in most of your code as they are easy to follow.

seantham21 commented on src/main/java/sage/storage/Storage.java:
```diff
@@ -0,0 +1,113 @@
+import sage.task.*;
+
+
+import java.io.*;
```
Perhaps more specific imports can be used?

seantham21 commented on src/main/java/sage/task/Event.java:
```diff
@@ -0,0 +1,49 @@
+    @Override
+    public String toFileFormat() {
+        DateTimeFormatter fileFormatter = DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm");
+        return "E | " + (isDone ? "1" : "0") + " | " + description + " | " + from.format(fileFormatter) + " | " + to.format(fileFormatter);
```
Perhaps a line split could be used at an appropriate place to prevent this long line?

seantham21 reviewed (commented):
Overall, just some minor changes requested for an even better readability of code. 
LGTM nonetheless so good job!

jiayr99 committed:
Add jar file

jiayr99 committed:
Add assertions to Storage and Parser class

There is an assertion in storage class to ensure that the file being processed is a file and not a directory.
There is another assertion in storage class to ensure task lines are correctly split into at least 3 parts.

There is an assertion in parser class to check that the user command is not null or blank before processing the command.
There is another assertion in parser class to ensure the input is not empty.

jiayr99 committed:
Improve code quality

jiayr99 committed:
Merge pull request #1 from jiayr99/branch-A-Assertions

Add assertions to Storage and Parser class

jiayr99 committed:
Merge pull request #2 from jiayr99/branch-A-CodeQuality

Improve code quality

jiayr99 committed:
Add a command to view tasks in the form of a schedule

jiayr99 committed:
Improve on GUI

jiayr99 committed:
Merge branch 'branch-A-BetterGui'

jiayr99 committed:
Give the bot a personality

jiayr99 committed:
Merge branch 'branch-A-Personality'

jiayr99 committed:
Add more error handling

jiayr99 committed:
Merge branch 'branch-A-MoreErrorHandling'

jiayr99 committed:
Add more testing

jiayr99 committed:
Merge branch 'branch-A-MoreTesting'

jiayr99 committed:
Update README.md

jiayr99 committed:
Update README.md

jiayr99 committed:
Revert "Update README.md"

This reverts commit ac5896c1c92d42e38f67f3e08d984d38965a4320.

jiayr99 committed:
Merge branch 'master' of https://github.com/jiayr99/ip
